pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -0,0 +1,277 @@
+		shards, err := strconv.ParseInt(value, 10, 32)
+		if err != nil {
+			continue
+		}
+		if shards > 0 {
+			spanLog.DebugLog(
+				"msg",
+				fmt.Sprintf("using shard count from header %s: %d", totalShardsControlHeader, shards),
+			)
+			return int(shards)
+		}
+	}
+	return origShardCount
+}
+
+func buildShardedRequests(ctx context.Context, req *http.Request, numRequests int, selector parser.Expr) ([]*http.Request, error) {
+	reqs := make([]*http.Request, numRequests)
+	for i := 0; i < numRequests; i++ {
+		reqs[i] = req.Clone(ctx)
+
+		sharded, err := shardedSelector(numRequests, i, selector)
+		if err != nil {
+			return nil, err
+		}
+
+		vals := url.Values{}
+		vals.Set("selector", sharded.String())
+
+		reqs[i].Header.Set("Content-Type", "application/x-www-form-urlencoded")
+		reqs[i].Header.Del(totalShardsControlHeader)
+		reqs[i].Body = io.NopCloser(strings.NewReader(vals.Encode()))
+	}
+
+	return reqs, nil
+}
+
+func doShardedRequests(ctx context.Context, upstreamRequests []*http.Request, next http.RoundTripper) ([]*http.Response, error) {
+	mtx := sync.Mutex{}
+	var resps []*http.Response
+
+	g, ctx := errgroup.WithContext(ctx)
+	queryStats := stats.FromContext(ctx)
+	for _, req := range upstreamRequests {
+		r := req
+		g.Go(func() error {
+			partialStats, childCtx := stats.ContextWithEmptyStats(ctx)
+			partialStats.AddShardedQueries(1)
+
+			var span opentracing.Span
+			span, childCtx = opentracing.StartSpanFromContext(childCtx, "shardActiveSeries.doRequests")
+			defer span.Finish()
+
+			resp, err := next.RoundTrip(r.WithContext(childCtx))
+			if err != nil {
+				span.LogFields(otlog.Error(err))
+				return err
+			}
+
+			queryStats.Merge(partialStats)
+
+			mtx.Lock()
+			resps = append(resps, resp)
+			mtx.Unlock()
+
+			return nil
+		})
+	}
+
+	return resps, g.Wait()
+}
+
+func shardedSelector(shardCount, currentShard int, expr parser.Expr) (parser.Expr, error) {
+	originalSelector, ok := expr.(*parser.VectorSelector)
+	if !ok {
+		return nil, errors.New("invalid selector")
+	}
+
+	shardMatcher, err := labels.NewMatcher(
+		labels.MatchEqual, sharding.ShardLabel,
+		sharding.ShardSelector{ShardIndex: uint64(currentShard), ShardCount: uint64(shardCount)}.LabelValue(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &parser.VectorSelector{
+		Name:          originalSelector.Name,
+		LabelMatchers: append([]*labels.Matcher{shardMatcher}, originalSelector.LabelMatchers...),
+	}, nil
+}
+
+func (s *shardActiveSeriesMiddleware) mergeResponses(responses []*http.Response) *http.Response {
+	reader, writer := io.Pipe()
+
+	json = jsoniter.ConfigCompatibleWithStandardLibrary
+	items := make(chan any)
+
+	g := new(errgroup.Group)
+	for _, res := range responses {
+		r := res
+		g.Go(func() error {
+			defer func(Body io.ReadCloser) {
+				_ = Body.Close()
+			}(r.Body)
+
+			it := jsoniter.Parse(jsoniter.ConfigFastest, r.Body, 512)
+
+			field := it.ReadObject()
+			if field != "data" {
```

Comment:
I hope nobody ever adds a `success` field to `activeSeriesResponse`. This is quite flaky.

Reply:
I made this more robust by iterating through the top level fields.